storage/keyring.go

```diff
@@ -165,34 +166,105 @@ var storageDataUnmarshal Unmarshaler = json.Unmarshal
 
 // loads the entire StorageData into memory
 func (kr *KeyringStore) getStorageData(s *StorageData) error {
-	data, err := kr.keyring.Get(RECORD_KEY)
+	var data []byte
+	var err error
+
+	switch getMyGOOS() {
+	case "windows":
+		data, err = kr.joinAndGetKeyringData(RECORD_KEY)
+	default:
+		data, err = kr.getKeyringData(RECORD_KEY)
+	}
+
 	if err != nil {
 		// Didn't find anything in our keyring
 		*s = NewStorageData()
 		return nil
 	}
-	if err = storageDataUnmarshal(data.Data, s); err != nil {
+	if err = storageDataUnmarshal(data, s); err != nil {
 		return err
 	}
 	return nil
 }
 
+// reads a single entry of the keyring
+func (kr *KeyringStore) getKeyringData(key string) ([]byte, error) {
+	data, err := kr.keyring.Get(key)
+	if err != nil {
+		return nil, err
+	}
+	return data.Data, nil
+}
+
+// read the chunks stored in windows credential manager
+func (kr *KeyringStore) joinAndGetKeyringData(key string) ([]byte, error) {
+	i := 0
+	data := []byte{}
+	var err error
+	for {
```

Comment:
This function breaks the contract that is implicit when calling getKeyringData().  Ie: if you don't get an error, you are guaranteed to have the full and complete set of data for parsing.  Since this is now done in chunks and there is no indicator what the total length of data that should be read, there is no way to know if you have everything unless until you try parsing.

Basically, what happens when there are 3 chunks and someone mistakenly deletes the 2nd or 3rd chunk?  Instead of returning a short read, we should return an error.

I'd suggest to store the number of bytes as a uint64 as the first 8 bytes then you can verify that the data read is complete.

Reply:
Ok, let me see if I get all your suggestions. 
```golang
// read the chunks stored in windows credential manager
func (kr *KeyringStore) joinAndGetKeyringData(key string) ([]byte, error) {
	i := 0
	data := []byte{}
	var err error
	var payload_size uint64

	for {
		var chunk []byte
		chunk, err = kr.getKeyringData(fmt.Sprintf("%s_%d", key, i))
		if i == 0 {
			if len(chunk) < 8 {
				return nil, fmt.Errorf("Invalid stored data in wincred. There's not enough data")
			}
			payload_size, chunk = getUINT64FromBytes(chunk[:8]), chunk[8:]
		}
		if chunk != nil && len(chunk) <= WINCRED_MAX_LENGTH {
			data = append(data, chunk...)
		}
		if err != nil || len(chunk) < WINCRED_MAX_LENGTH {
			break
		}
		i++
	}

	if i == 0 && err != nil {
		return nil, err
	}

	if payload_size != uint64(len(data)) {
		return nil, fmt.Errorf("Invalid stored data in wincred")
	}
	return data, nil
}
```

Reply:
fyi, there is no need to copy & paste the code in the comments and it actually makes it harder to review.